Modification of 'Adding modules and plugins locally' topic (#70659)
ansible/ansible: docs/docsite/rst/dev_guide/developing_locally.rst

Adding modules and plugins locally
**********************************

> The easiest, quickest, and the most popular way to extend Ansible is to use a module or a plugin. You can create them or copy existing ones for local use. You can store a local module or plugin on your Ansible control node and share it with your team or organization. You can also share a local plugin or module by embedding it in a role and publishing it on Ansible Galaxy. If you are using roles on Ansible Galaxy, then you are already using local modules and plugins without realizing it.

Reviewer: The first sentence here, as changed, is not quite accurate. Modules and plugins are the building blocks of Ansible behavior. This page shows users how to get a **local** plugin or module to work.
```suggestion
The easiest, quickest, and the most popular way to extend Ansible is to use a local module or a plugin. You can create them or copy existing ones for local use. You can store a local module or plugin on your Ansible control node and share it with your team or organization. You can also share a local plugin or module by embedding it in a role and publishing it on Ansible Galaxy. If you are using roles on Ansible Galaxy, then you are already using local modules and plugins without realizing it.
```

If you are using an existing module or plugin but Ansible can't find it, this page is all you need. However, if you want to create a plugin or a module, go to :ref:`developing_plugins` and :ref:`developing_modules_general` topics and then return to this page to know how to add it locally. 

Extending Ansible with local modules and plugins offers lots of shortcuts such as:

* You can copy other people's modules and plugins.
* If you're writing a new module, you can choose any programming language you like.
* You don't have to clone the main Ansible repo.
* You don't have to open a pull request.
* You don't have to add tests (though we recommend that you do!).

To save a local module or plugin such that Ansible can find and use it, add the module or plugin in the appropriate directory (the directories are specified in later parts of this topic). 

.. contents::
   :local:

.. _modules_vs_plugins:

> What are modules and plugins?

Reviewer 2: I would keep the original one.

Author: Thank you for the review, @Akasurde. I updated the title in the commit instance 0880239.

===========================================
If you're looking to add local functionality to Ansible, you might wonder whether you need a module or a plugin. Here's a quick overview to help you decide between the two:

> * Modules are reusable, standalone scripts that can be used by the Ansible API, the :command:`ansible` command, or the :command:`ansible-playbook` command. Modules provide a defined interface, accept arguments, and return information to Ansible by printing a JSON string to stdout before exiting. Modules execute on the target system (usually that means on a remote system) in separate processes.

Reviewer: This change altered the meaning slightly, but the sentence could definitely be clarified. How about:
```suggestion
* Modules are reusable, standalone scripts that can be used by the Ansible API, the :command:`ansible` command, or the :command:`ansible-playbook` command. Modules provide a defined interface. Each module accepts arguments and returns information to Ansible by printing a JSON string to stdout before exiting. Modules execute on the target system (usually that means on a remote system) in separate processes.
```

* :ref:`Plugins <plugins_lookup>` augment Ansible's core functionality and execute on the control node within the ``/usr/bin/ansible`` process. Plugins offer options and extensions for the core features of Ansible - transforming data, logging output, connecting to inventory, and more.

.. _local_modules:

Adding a module locally
=======================
Ansible automatically loads all executable files found in certain directories as modules.

For local modules, use the name of the file as the module name: for example, if the module file is ``~/.ansible/plugins/modules/local_users.py``, use ``local_users`` as the module name.

> To load your local modules automatically, add them in any of these locations:

Reviewer: ```suggestion
To load your local modules automatically and make them available to all playbooks and roles, add them in any of these locations:
```

.. note::

> 	Currently, the ``ansible-doc`` command can parse Python modules only for the module documentation. If you have a module written in a different programming language other than Python, please write the documentation in a Python file adjacent to the module file.

Reviewer: Nice catch here. This section needed revision. The placement of `only` in the revised version feels awkward, though. How about this:
 ```suggestion
	Currently, the ``ansible-doc`` command can parse module documentation only from modules written in Python. If you have a module written in a programming language other than Python, please write the documentation in a Python file adjacent to the module file.
```

> The location of your module depends on whether you want to use it in a playbook or in a single role. Choose any of the following locations to add your module: 

Reviewer: This list is about limiting the scope/context for a local module. 
```suggestion
You can limit the availability of your local module. If you want to use a local module only with selected playbooks or only with a single role, load it in one of the following locations:
```

> * In a certain playbook: Store the module in a subdirectory called ``library`` in the directory that contains the playbooks.

Reviewer: ```suggestion
* In a selected playbook or playbooks: Store the module in a subdirectory called ``library`` in the directory that contains those playbooks.
```

    * test_plugins*
    * vars_plugins

.. note::

	After you add the plugins and verify that they are available for use, you can see the documentation for all the plugins except for the ones marked with an asterisk (*) above.

To load your local plugins automatically, add them in any of these locations:

* any directory added to the relevant ``ANSIBLE_plugin_type_PLUGINS`` environment variable (these variables, such as ``$ANSIBLE_INVENTORY_PLUGINS`` and ``$ANSIBLE_VARS_PLUGINS`` take colon-separated lists like ``$PATH``)
* the directory named for the correct ``plugin_type`` within ``~/.ansible/plugins/`` - for example, ``~/.ansible/plugins/callback``
* the directory named for the correct ``plugin_type`` within ``/usr/share/ansible/plugins/`` - for example, ``/usr/share/ansible/plugins/action``

> After your plugin file is in one of these locations, Ansible loads it and you can use it in any local module, task, playbook, or role. Alternatively, you can edit your ``ansible.cfg`` file to add directories that contain local plugins. For information about adding directories of local plugins, see :ref:`ansible_configuration_settings` for details.

Reviewer: ```suggestion
After your plugin file is in one of these locations, Ansible loads it and you can use it in any local module, task, playbook, or role. Alternatively, you can edit your ``ansible.cfg`` file to add directories that contain local plugins. For details about adding directories of local plugins, see :ref:`ansible_configuration_settings`.
```

To confirm that ``plugins/plugin_type/my_custom_plugin`` is available:

* type ``ansible-doc -t <plugin_type> my_custom_lookup_plugin``. For example, ``ansible-doc -t lookup my_custom_lookup_plugin``. You should see the documentation for that plugin. This works for all plugin types except the ones marked with ``*`` in the list above  - see :ref:`ansible-doc` for more details.

> The location of your plugin depends on whether you want to use it in a playbook or in a single role.  Choose any of the following locations to add your plugin:

Reviewer: Same comment here as above - this is about limiting the scope of a plugin.
```suggestion
You can limit the availability of your local plugin. If you want to use a local plugin only with selected playbooks or only with a single role, load it in one of the following locations:
```

> * In a certain playbook: Store the plugin in a subdirectory for the correct ``plugin_type`` (for example, ``callback_plugins`` or ``inventory_plugins``) in the directory that contains the playbooks.

Reviewer: ```suggestion
* In a selected playbook or playbooks: Store the plugin in a subdirectory for the correct ``plugin_type`` (for example, ``callback_plugins`` or ``inventory_plugins``) in the directory that contains the playbooks.
```
